fdbclient/include/fdbclient/CommitTransaction.h

```diff
@@ -123,59 +129,124 @@ struct MutationRef {
+		static_assert(MAX_ATOMIC_OP < 64U); // 2 bits have been reserved for checksum and accumulative checksum index
+		return type < MAX_ATOMIC_OP;
+	}
+
+	// If mutation checksum is enabled, we must set accumulative index before serialization
+	// Once accumulative index is set, it cannot change over time
+	void setAccumulativeChecksumIndex(uint16_t index) {
+		if (this->accumulativeChecksumIndex.present()) {
+			ASSERT(this->accumulativeChecksumIndex.get() == index);
+		} else {
+			this->accumulativeChecksumIndex = index;
 		}
-		ASSERT(this->param2.size() >= 4);
-		const int idx = this->param2.size() - 4;
-		const uint32_t pc = *(const uint32_t*)(this->param2.substr(idx).begin());
+		return;
+	}
+
+	// If param2 includes checksum suffix, remove the suffix and set it to this->checksum
+	// Unflag the corresponding bit in type
+	// This operation must be after removing the acs index if exists
+	void offloadChecksum() {
+		ASSERT(!this->checksum.present());
+		if (!this->withChecksum()) {
+			return;
+		}
+		ASSERT(!this->withAccumulativeChecksumIndex()); // Removing checksum must be after removing acs index
 		this->type &= ~CHECKSUM_FLAG_MASK;
-		this->param2 = this->param2.substr(0, idx);
-		return pc;
+		int index = this->param2.size() - 4;
+		this->checksum = *(const uint32_t*)(this->param2.substr(index, 4).begin());
+		this->param2 = this->param2.substr(0, index);
 	}
-	bool withChecksum() const { return this->type & CHECKSUM_FLAG_MASK; }
 
-	bool isAtomicOp() const { return (ATOMIC_MASK & (1 << type)) != 0; }
-	bool isValid() const { return type < MAX_ATOMIC_OP; }
+	// If param2 includes acs index suffix, remove the suffix and set it to this->accumulativeChecksumIndex
+	// Unflag the corresponding bit in type
+	void offloadAccumulativeChecksumIndex() {
+		ASSERT(!this->accumulativeChecksumIndex.present());
+		if (!this->withAccumulativeChecksumIndex()) {
+			return;
+		}
+		this->type &= ~ACCUMULATIVE_CHECKSUM_INDEX_FLAG_MASK;
+		int index = this->param2.size() - 2;
+		this->accumulativeChecksumIndex = *(const uint16_t*)(this->param2.substr(index, 2).begin());
+		this->param2 = this->param2.substr(0, index);
+	}
 
-	uint32_t populateChecksum() {
-		ASSERT(!this->withChecksum());
-		uint32_t c = crc32c_append(static_cast<uint32_t>(this->type), param1.begin(), param1.size());
-		crc32c_append(c, param2.begin(), param2.size());
-		if (this->checksum.present()) {
-			if (this->checksum.get() != c) {
-				TraceEvent(SevError, "MutationRefChecksumMismatch")
-				    .detail("CalculatedChecksum", std::to_string(c))
-				    .detail("Mutatino", toString());
-			}
+	// Clear checksum and acs index if exist
+	void clearChecksumAndAccumulativeIndex() {
+		this->checksum.reset();
+		this->accumulativeChecksumIndex.reset();
+	}
+
+	// Validate param2 correctness
+	void validateParam2() {
+		if (this->withChecksum() && this->withAccumulativeChecksumIndex()) {
+			ASSERT(this->param2.size() >= 6); // 4 bytes for checksum, 2 bytes for accumulative index
```

Comment:
Are there other actions we can take in case of a corruption, since it could result in an infinite crash loop.

Discussed offline with Zhe, we will handle this more gracefully.